state/raftlease/target.go

```diff
@@ -174,33 +176,110 @@ func (t *notifyTarget) Claimed(key lease.Key, holder string) error {
 	return errors.Annotatef(err, "%q for %q in db", docId, holder)
 }
 
-// Expired is part of raftlease.NotifyTarget.
-func (t *notifyTarget) Expired(key lease.Key) error {
+type leaseDoc struct {
+	Key    lease.Key
+	DocId  string
+	Holder string
+}
+
+// Expiries is part of raftlease.NotifyTarget.
+func (t *notifyTarget) Expiries(expiries []raftlease.Expired) error {
+	if len(expiries) == 0 {
+		return nil
+	}
+
 	coll, closer := t.mongo.GetCollection(t.collection)
 	defer closer()
 
-	docId := leaseHolderDocId(key.Namespace, key.ModelUUID, key.Lease)
-	t.logger.Infof("expiring lease %q", docId)
+	// Cache all the document idents up front, incase we have to retry the
+	// transaction again. Also this serves as a de-duping process.
+	leaseDocs := make(map[string]leaseDoc)
+	for _, expired := range expiries {
+		key := expired.Key
+		docId := leaseHolderDocId(key.Namespace, key.ModelUUID, key.Lease)
+
+		// We have an existing lease.Key already in the documents that doesn't
+		// have the same holder.
```

Comment:
This comment should either be "Do we have an existing", or it should be inside the if block, so that it is clear that we have determined this to be true.